### DISCUSSION TITLE
Use as go package

### DISCUSSION THREAD

**Discussion by codegoalie** [jotaen/klog#373] [Help / Question]
First off, thanks so much for creating klog! I've been using it for a while now just to track my work time. It's so great to have something easy and predictable that doesn't need a network connection! 

I recently started doing some consulting work and want to programatically convert by klog entries into an invoice. I would lke to import some of the parsing features into an existing go program. However, it seems by declaring the base package as just `klog`, the go tooling can't find the code to download it. It seems changing to `github.com/jotaen/klog` would work. Is that a reasonable change? Would that cause more problems than it's worth? Thanks!

**Comment by jotaen** [comment]
Glad to hear that you enjoy klog.

I created [this branch](https://github.com/jotaen/klog/compare/packaging?expand=1) to see what needs changing to make your suggestion happen. You can try it by running `go get github.com/jotaen/klog@fac2e99` and then importing via `require "github.com/jotaen/klog/src"`. Could you give this a shot and see whether it works on your end too?

I cannot think of any technical issues when renaming the package. Actually, I suppose your suggestion is more common practice and in line with the “standard” of go modules. My concern is more about usage: this repo is primarily meant for building the CLI tool, and not for being used as standalone source code library. The reason is that I don’t want to go down the road of creating proper API documentation, making compatibility guarantees, or providing code-level support. That’d be a bit too much for a spare time project, unfortunately.

So, as long as it’s clear that people are “on their own” when using the sources directly, I’m happy to make that process more convenient. I still need to think some more about this before merging, but I’m generally positive (with the above caveat, though). I’d also need to add a sentence or two to the README then.

**Comment by codegoalie** [comment]
Sorry it took me so long to get back to this. But, it worked!! 

> I don’t want to go down the road of creating proper API documentation, making compatibility guarantees, or providing code-level support

Totally understand! I think a few lines to the README wouldn't be out of line. I was (and still am) proceeding with the understanding that I was peering under the hood and to proceed at my own risk. 

The one small thing I wanted from klog that started me down this path was fractional hours of the entry durations. E.g. `14:00 - 14:30 Consult with client` to be printed out or represented like: "0.5 Consult with client". Then I could more easily copy/paste onto the invoice.

Once I started looking into the code, I noticed how you had things nicely decoupled and the parser was separate. Seeing that it was one method call, `parser.Parse`, to get a list of `Record`s from a file path got me thinking more ambitiously about completely automating the invoice generation process.

Looking at the branch, the change is mostly to import paths. Not too too bad.

Anyway, thank you so so much!! I'm super stoked about this! Let me know if I can help in any way.

**Comment by jotaen** [comment]
Awesome, thanks for confirming! I’ll wrap up the branch probably next week or so.

Just for the records: there is also the `klog json` command, which converts klog data into a JSON structure. (But if you are using Go already for your app, it’s probably easier (and more powerful) to integrate directly against the code.)

**Comment by codegoalie** [comment]
Oh man. How did I miss `klog json`? 🤦🏻 That _may_ have worked, but integrating directly in Go is a little better for me. Thanks again!!

**Comment by jotaen** [comment]
Done – thanks @codegoalie for the idea!

Just for future reference: keep in mind to always pin the commit hash, as there won’t be separate releases of the source code. E.g.:

```bash
go get github.com/jotaen/klog@07ba740
```

(This would pull the klog source code at commit 07ba740. If you omit the commit hash, then it falls back to the latest commit, which might break your code at some point in the future.)

**Comment by codegoalie** [comment]
Amazing! Thank you so so much!
